presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapSslSocketFactory.java

```diff
@@ -0,0 +1,108 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.password.ldap;
+
+import javax.net.SocketFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+
+public class LdapSslSocketFactory
+        extends SSLSocketFactory
+{
+    private static final ThreadLocal<SSLContext> SSL_CONTEXT = new ThreadLocal<>();
+
+    private final SSLSocketFactory socketFactory;
+
+    public LdapSslSocketFactory(SSLContext sslContext)
+    {
+        this.socketFactory = sslContext.getSocketFactory();
+    }
+
+    @Override
+    public String[] getDefaultCipherSuites()
+    {
+        return socketFactory.getDefaultCipherSuites();
+    }
+
+    @Override
+    public String[] getSupportedCipherSuites()
+    {
+        return socketFactory.getSupportedCipherSuites();
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, String host, int port, boolean autoClose)
+            throws IOException
+    {
+        return socketFactory.createSocket(socket, host, port, autoClose);
+    }
+
+    @Override
+    public Socket createSocket(String host, int port)
+            throws IOException
+    {
+        return socketFactory.createSocket(host, port);
+    }
+
+    @Override
+    public Socket createSocket(String host, int port, InetAddress clientAddress, int clientPort)
+            throws IOException
+    {
+        return socketFactory.createSocket(host, port, clientAddress, clientPort);
+    }
+
+    @Override
+    public Socket createSocket(InetAddress address, int port)
+            throws IOException
+    {
+        return socketFactory.createSocket(address, port);
+    }
+
+    @Override
+    public Socket createSocket(InetAddress address, int port, InetAddress clientAddress, int clientPort)
+            throws IOException
+    {
+        return socketFactory.createSocket(address, port, clientAddress, clientPort);
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, InputStream consumed, boolean autoClose)
+            throws IOException
+    {
+        return socketFactory.createSocket(socket, consumed, autoClose);
+    }
+
+    @Override
+    public Socket createSocket()
+            throws IOException
+    {
+        return socketFactory.createSocket();
+    }
+
+    @SuppressWarnings("MethodOverridesStaticMethodOfSuperclass")
+    public static SocketFactory getDefault()
+    {
+        return new LdapSslSocketFactory(SSL_CONTEXT.get());
```

Comment:
verify SSL_CONTEXT.get() is not null

Reply:
you delegate all [explicitly implemented] methods

why not just 
```
return SSL_CONTEXT.get().getSocketFactory();
```
?

Reply:
Ooh, that's a nice trick.

Reply:
It turns out that in Java 8, the JDK code assumes that `createSocket` is implemented on the passed in class name (it looks it up reflectively), and that the factory returned from `getDefault()` is of the same class. The Java 11 code works as expected by casting to `SocketFactory` and simply calling the method.

Reply:
Thanks for checking this! Please add a TODO so we can prune the boilerplate when we drop J 8